jbmc/src/java_bytecode/java_trace_verification.cpp

```diff
@@ -0,0 +1,181 @@
+}
+
+bool check_member_structure(const exprt &member_expr)
+{
+  if(!expr_try_dynamic_cast<member_exprt>(member_expr))
+    return false;
+  if(!member_expr.has_operands())
+    return false;
+  const auto symbol_operand = get_inner_symbolexpr(member_expr);
+  if(!symbol_operand)
+    return false;
+  return true;
+}
+
+static void print_error(const std::string &side, const exprt &expr)
+{
+  throw std::runtime_error(
+    "JBMC Check trace assumption failure on " + side + " expression:\n" +
+    expr.pretty(0, 0));
+}
+
+bool valid_lhs_expr_high_level(const exprt &lhs)
+{
+  return expr_try_dynamic_cast<member_exprt>(lhs) ||
+         expr_try_dynamic_cast<symbol_exprt>(lhs) ||
+         expr_try_dynamic_cast<index_exprt>(lhs) ||
+         lhs.id() == ID_byte_extract_little_endian;
+}
+
+bool valid_rhs_expr_high_level(const exprt &rhs)
+{
+  return expr_try_dynamic_cast<struct_exprt>(rhs) ||
+         expr_try_dynamic_cast<array_exprt>(rhs) ||
+         expr_try_dynamic_cast<constant_exprt>(rhs) ||
+         expr_try_dynamic_cast<address_of_exprt>(rhs) ||
+         expr_try_dynamic_cast<symbol_exprt>(rhs) ||
+         expr_try_dynamic_cast<array_list_exprt>(rhs);
+}
+
+void check_trace_assumptions(const goto_tracet &trace, const namespacet &ns)
+{
+  for(const auto &step : trace.steps)
+  {
+    if(!step.is_assignment() && !step.is_decl())
+      continue;
+    const auto &lhs = skip_typecast(step.full_lhs);
+    const auto &rhs = skip_typecast(step.full_lhs_value);
+    if(!valid_lhs_expr_high_level(lhs))
+      print_error("LHS", lhs);
+    // check member lhs structure
+    if(const auto member = expr_try_dynamic_cast<member_exprt>(lhs))
+    {
+      if(!check_member_structure(*member))
+        print_error("LHS", lhs);
+    }
+    // check symbol lhs structure
+    if(const auto symbol = expr_try_dynamic_cast<symbol_exprt>(lhs))
+    {
+      if(!check_symbol_structure(*symbol))
+        print_error("LHS", lhs);
+    }
+    // check index lhs structure
+    if(const auto index = expr_try_dynamic_cast<index_exprt>(lhs))
+    {
+      if(index->operands().size() != 2)
+        print_error("LHS", lhs);
+      if(!check_symbol_structure(index->op0()))
+        print_error("LHS", lhs);
+      if(!expr_try_dynamic_cast<constant_exprt>(index->op1()))
+        print_error("LHS", lhs);
+    }
+    // check byte extract lhs structure
+    if(lhs.id() == ID_byte_extract_little_endian)
+    {
+      if(!check_symbol_structure(lhs.op0()))
+        print_error("LHS", lhs);
+      if(!expr_try_dynamic_cast<constant_exprt>(simplify_expr(lhs.op1(), ns)))
+        print_error("LHS", lhs);
+    }
+
+    if(!valid_rhs_expr_high_level(rhs))
+      print_error("RHS", rhs);
+    // check address_of rhs structure (String only)
+    if(const auto address = expr_try_dynamic_cast<address_of_exprt>(rhs))
+    {
+      const auto symbol_expr = get_inner_symbolexpr(*address);
+      if(!symbol_expr)
+        print_error("RHS", rhs);
```

Comment:
:pick: it would be good to have more detailed error messages, like "only address of symbols are allowed"

Reply:
I think this might make it quite hard to maintain.

Reply:
I don't see why?

Reply:
I've added them, but I have a feeling that the functions and these messages will get out of sync.